numba/cpython/tupleobj.py

```diff
@@ -380,3 +383,156 @@ def tuple_index_impl(tup, value):
+                "make_tuple: arrays provided as 'iterable' must be 1d")
+    if hasattr(iterable, 'dtype'):
+        dtype = iterable.dtype
+    elif hasattr(iterable, 'yield_type'):
+        dtype = iterable.yield_type
+    elif hasattr(iterable, 'key_type'):
+        dtype = iterable.key_type
+    elif hasattr(iterable, 'item_type'):
+        dtype = iterable.item_type
+    elif hasattr(iterable, 'literal_type'):
+        dtype = iterable.literal_type
+    elif isinstance(iterable, types.UnicodeType):
+        dtype = iterable
+    else:
+        raise TypeError(
+            "make_tuple: cannot determine dtype of argument 'iterable'"
+            f" of type '{iterable}'")
+
+    tuple_type = types.UniTuple(dtype, tuple_size)
+    sig = tuple_type(size, iterable)
+
+    # Separate implementations are required to ensure that for iterators
+    # multiple calls continue the iteration while multiple calls
+    # with array, lists, etc must restart at the beginning
+    if isinstance(iterable, (types.IteratorType, types.SimpleIterableType, 
+                             types.DictType)):
+        def convert_to_list(iterable):
+            i = 0
+            out = typed.List()
+            for i, val in enumerate(iterable):
+                out.append(val)
+                if i+1 == tuple_size:
+                    break
+            if i+1 != tuple_size:
+                raise ValueError(too_short_msg)
+            return out
+    else:
+        def convert_to_list(iterable):
+            if len(iterable) < tuple_size:
+                raise ValueError(too_short_msg)
+            out = typed.List()
+            for i in range(tuple_size):
+                out.append(iterable[i])
+            return out
+
+    def codegen(context, builder, signature, args):
+        [_, iterable_ty] = sig.args
+        [_, iterable] = args
+
+        # Normalize the input into a list with the requested number of elements
+        inner_argtypes = [iterable_ty]
+        inner_rettype = types.ListType(dtype)
+        inner_sig = typing.signature(inner_rettype, *inner_argtypes)
+        inner_args = [iterable]
+        tmp_list = context.compile_internal(
+            builder, convert_to_list, inner_sig, inner_args)
+
+        # Allocate an empty tuple
+        out_tuple = context.get_constant_null(tuple_type)
+
+        # Copy the list elements into the tuple
+        def list_indexer(a, i):
+            return a[i]
+
+        for i in range(tuple_size):
+            list_idx = context.get_constant(types.intp, i)
+            data = context.compile_internal(builder, list_indexer,
+                                          dtype(inner_rettype, types.intp),
+                                          [tmp_list, list_idx])
+            out_tuple = builder.insert_value(out_tuple, data, i)
+
+        # Clean up the temporary list
+        if context.enable_nrt:
+            context.nrt.decref(builder, inner_rettype, tmp_list)
+
+        return out_tuple
```

Comment:
I'm wondering if this approach could be simplified. At compile time the size of the tuple is known, so can be allocated via an `@intrinsic`, subsequently all writes can be done via an intrinsic too, and the driving code be done in a `njit` function? I'm struggling to understand the need for a list buffer too?